source/extensions/formatter/xfcc_value/xfcc_value.cc

```diff
@@ -0,0 +1,204 @@
+                                                               "by",
+                                                               "hash",
+                                                               "cert",
+                                                               "chain",
+                                                               "subject",
+                                                               "uri",
+                                                               "dns",
+                                                           });
+}
+
+// The XFCC header value is a comma (`,`) separated string. Each substring is an XFCC element,
+// which holds information added by a single proxy. A proxy can append the current client
+// certificate information as an XFCC element, to the end of the request’s XFCC header after a
+// comma.
+//
+// Each XFCC element is a semicolon (`;`) separated string. Each substring is a key-value pair,
+// grouped together by an equals (`=`) sign. The keys are case-insensitive, the values are
+// case-sensitive. If  `,`, `;` or `=` appear in a value, the value should be double-quoted.
+// Double-quotes in the value should be replaced by backslash-double-quote (`\"`).
+// Only need to find the first value with the key, so scan the header left-to-right.
+
+// Handles a single key/value pair within an XFCC element.
+absl::optional<std::string> parseKeyValuePair(absl::string_view pair, absl::string_view target) {
+  // Find '=' not in quotes. Because key will always be the first part and won't be double
+  // quoted or contain `=`, we can safely use `absl::StrSplit`.
+  std::pair<absl::string_view, absl::string_view> key_value =
+      absl::StrSplit(pair, absl::MaxSplits('=', 1));
+  absl::string_view raw_key = absl::StripAsciiWhitespace(key_value.first);
+  if (!absl::EndsWithIgnoreCase(raw_key, target)) {
+    return absl::nullopt;
+  }
+
+  absl::string_view raw_value = absl::StripAsciiWhitespace(key_value.second);
+  // If value is double quoted, remove quotes.
+  if (raw_value.size() >= 2 && raw_value.front() == '"' && raw_value.back() == '"') {
+    raw_value = raw_value.substr(1, raw_value.size() - 2);
+  }
+
+  // Unescape double quotes.
+  // Note: this only handle escaped double quotes, not other escape sequences.
+  std::string unescaped;
+  for (size_t i = 0; i < raw_value.size(); ++i) {
+    if (raw_value[i] == '\\' && i + 1 < raw_value.size() && raw_value[i + 1] == '"') {
+      unescaped.push_back('"');
+      ++i;
+    } else {
+      unescaped.push_back(raw_value[i]);
+    }
+  }
+  return unescaped;
+}
+
+// Handles a single XFCC element (semicolon-separated key/value pairs).
+absl::optional<std::string> parseElementForKey(absl::string_view element,
+                                               absl::string_view target) {
+
+  // Scan key-value pairs in this element (by semicolon not in quotes).
+  bool in_quotes = false;
```

Comment:
A hand-written parser makes me a bit worried. I understand this is a formatter, but it seems like the intent is to use for security features (e.g. authz by cert fields), so we should make sure this is well tested (and maybe fuzzed).

Reply:
Ack, a fuzzer should be enough, plus a careful code review. It's not a particularly hard grammar.